core/src/main/java/ai/timefold/solver/core/impl/move/streams/maybeapi/generic/provider/ListSwapMoveProvider.java

```diff
@@ -0,0 +1,65 @@
+package ai.timefold.solver.core.impl.move.streams.maybeapi.generic.provider;
+
+import java.util.Objects;
+
+import ai.timefold.solver.core.impl.move.streams.maybeapi.DataJoiners;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.generic.move.ListSwapMove;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.stream.MoveProducer;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.stream.MoveProvider;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.stream.MoveStreamFactory;
+import ai.timefold.solver.core.preview.api.domain.metamodel.PlanningListVariableMetaModel;
+import ai.timefold.solver.core.preview.api.domain.metamodel.PositionInList;
+import ai.timefold.solver.core.preview.api.move.SolutionView;
+
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public class ListSwapMoveProvider<Solution_, Entity_, Value_>
+        implements MoveProvider<Solution_> {
+
+    private final PlanningListVariableMetaModel<Solution_, Entity_, Value_> variableMetaModel;
+
+    public ListSwapMoveProvider(PlanningListVariableMetaModel<Solution_, Entity_, Value_> variableMetaModel) {
+        this.variableMetaModel = Objects.requireNonNull(variableMetaModel);
+    }
+
+    @Override
+    public MoveProducer<Solution_> apply(MoveStreamFactory<Solution_> moveStreamFactory) {
+        var assignedValueStream = moveStreamFactory.enumerate(variableMetaModel.type(), false)
+                .filter((solutionView,
+                        value) -> solutionView.getPositionOf(variableMetaModel, value) instanceof PositionInList);
+        var validAssignedValuePairStream = assignedValueStream.join(assignedValueStream,
```

Comment:
CS includes a method for generating distinct pairs of values, and it seems like a good fit here. What do you think about adding something similar for MS?

Reply:
I'm actually thinking of _removing_ it from CS. :-) It does more harm than good, but that's for a longer conversation.

I have a suspicion that this code won't stay here for long anyway; it's brittle and it will need a better solution long-term. Good enough for the first preview.